R/CageData_Integration.R

```diff
@@ -0,0 +1,191 @@
+#' @param cds If you want to extend 5' leaders downstream, to catch uorfs going into cds, include it.
+addFirstCdsOnLeaderEnds = function(fiveUTRs, cds){
+
+  if(length(cds) == 0){
+    message("cds is empty, returning without using it.")
+    return(fiveUTRs)
+  }
+  if(is.null(names(cds))){
+    message("cds have no names, returning without using it.")
+    return(fiveUTRs)
+  }
+
+  if(tryCatch({length(cds[names(fiveUTRs)]) == 0},error = function(err) return(T))){ # <- check valid names
+    message("cds have no matching names, returning without using it.")
+    return(fiveUTRs)
+  }
+
+  cdsForUTRs <- cds[names(fiveUTRs)] # get only the ones we need
+
+  firstExons <- phead(cdsForUTRs, 1L) #select first in every, they must be sorted!
+  gr <- unlist(firstExons,use.names = F)
+  gr$cds_id <- NULL; gr$cds_name <- NULL; gr$exon_rank <- NULL
+  gr$exon_id <- NA;  gr$exon_name <- NA;  gr$exon_rank <- NA
+  grl = relist(gr,firstExons)
+  fiveUTRsWithCdsExons <- pc(fiveUTRs, grl) #  why length != fiveUTRs
+
+  return( reduce(fiveUTRsWithCdsExons) )
+}
+
+#' Extend first exon of each transcript with length specified
+#' @param fiveUTRs The 5' leader sequences as GRangesList
+#' @param extension The number of basses upstream to add on transcripts
+extendsTSSExons = function(fiveUTRs, extension = 1000){
+  fiveAsgr <- unlist(fiveUTRs)
+
+  if(is.null(fiveAsgr$exon_rank)) stop("fiveUTRs need column called exon_rank, see ?makeTranscriptDbFromGFF")
+  firstExons <- fiveAsgr[fiveAsgr$exon_rank == 1]
+
+  posIDs <- firstExons[strand(firstExons) == "+"]
+  minIDs <- firstExons[strand(firstExons)  == "-"]
+  promo <- promoters(firstExons, upstream=extension)
+
+  start(firstExons[names(posIDs)]) <- start(promo[names(posIDs)])
+  end(firstExons[names(minIDs)]) <- end(promo[names(minIDs)])
+
+  return(firstExons)
+}
+
+#' Find max peak for each transcript,
+#' returns as data.table, without names, but with index
+#' @param cageOverlaps The cageOverlaps between cage and extended 5' leaders
+#' @param filteredrawCageData The filtered raw cage-data used to reassign 5' leaders
+#' @importFrom data.table as.data.table
+findMaxPeaks = function(cageOverlaps, filteredrawCageData){
+  dt <- as.data.table(filteredrawCageData)
+  dt <- dt[from(cageOverlaps)]
+  dt$to <- to(cageOverlaps)
+
+  test <- dt[,max(score), by = to]
+  names(test) <- c("to","score")
+  res <-  merge(  test, dt)
+
+  return(res[!duplicated(res$to)]) #check this line!!!
+}
+
+
+
+#' Finds max peaks per trancsript from reads in the cagefile
+#' @param fiveUTRs The 5' leader sequences as GRangesList
+#' @param filePath The location of the cage-file
+#' @param extension The number of basses upstream to add on transcripts
+#' @param filterValue The number of counts(score) to filter on for a tss to pass as hit
+#' @param cageAsGR Alternative for filePath, if you have cage-data already loaded in R as GRanges
+findNewTSS = function(fiveUTRs, filePath = NULL, extension, filterValue, cageAsGR = NULL){
```

Comment:
it cant have two competing inputs, either signature (which S3 way) or do type on the input and then process